### ISSUE TITLE
Unable to enqueue when not opened

### ISSUE THREAD

**Issue by Crysitna** [pytorch/xla#902]
it shows the following errors after several batches training. Wondering if anyone know what's happending? Great thanks ahead! 

Using `pytorch-0.1` with prebuild VM Image and run the model on 8 xla devices
```
2019-08-04 09:46:55.157619: E tensorflow/compiler/xla/xla_client/tf_logging.cc:11] Check failed: session->session()->Run( feed_inputs, {}, {cached_node.operations[0]}, &outputs) == ::tensorflow::Status::OK() (Failed precondition: From /job:tpu_worker/replica:0/task:0:
Unable to enqueue when not opened, queue: [0000:00:06.0 PE0 C2 MC1 TN1 Queue HBM_WRITE]. State is: FAILED vs. OK)
*** Begin stack trace ***
        tensorflow::CurrentStackTrace[abi:cxx11]()
        xla::XrtComputationClient::ReleaseHandles(std::vector<xla::XrtComputationClient::DeviceHandle, std::allocator<xla::XrtComputationClient::DeviceHandle> >*, std::function<xla::XrtSession::CachedNode const& (xla::XrtSession*, tensorflow::Scope const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)> const&, xla::metrics::Metric*, xla::metrics::Counter*)
        xla::XrtComputationClient::HandleReleaser()
        xla::util::TriggeredTask::Runner()


        clone
*** End stack trace ***

terminate called after throwing an instance of 'std::runtime_error'
  what():  tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:952 : Check failed: session->session()->Run( feed_inputs, {}, {cached_node.operations[0]}, &outputs) == ::tensorflow::Status::OK() (Failed precondition: From /job:tpu_worker/replica:0/task:0:
Unable to enqueue when not opened, queue: [0000:00:06.0 PE0 C2 MC1 TN1 Queue HBM_WRITE]. State is: FAILED vs. OK)
*** Begin stack trace ***
        tensorflow::CurrentStackTrace[abi:cxx11]()
        xla::XrtComputationClient::ReleaseHandles(std::vector<xla::XrtComputationClient::DeviceHandle, std::allocator<xla::XrtComputationClient::DeviceHandle> >*, std::function<xla::XrtSession::CachedNode const& (xla::XrtSession*, tensorflow::Scope const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)> const&, xla::metrics::Metric*, xla::metrics::Counter*)
        xla::XrtComputationClient::HandleReleaser()
        xla::util::TriggeredTask::Runner()


        clone
*** End stack trace ***
```

**Comment by dlibenzi** [comment]
Is this consistently happening?
Does, say, MNIST (test/test_train_mnist.py) train correctly on the same user/TPU VM pair?

**Comment by Crysitna** [comment]
mnist doesn't show this problem tho. and neither training another model with a tiny dataset in this framework. 

when I change the model to use only one or two device, this problem seems gone, yet sometimes other problems comes: 
```
 W tensorflow/core/distributed_runtime/rpc/grpc_remote_master.cc:157] RPC failed with status = Unavailable: Socket closed and grpc_error_string = "{"created":"@1565009491.573537357","description":"Error received from peer","file":"external/grpc/src/core/lib/surface/call.cc","file_line":1039,"grpc_message":"Socket closed","grpc_status":14}", maybe retrying the RPC
```

**Comment by Crysitna** [comment]
update: wondering would that due to the large memory or long training time? I tried a smaller model with a smaller dataset and it works quite normally

**Comment by dlibenzi** [comment]
Are you able to switch to use nightly instead of r0.1? Nightly gets the latest changes.

So single core works for you, with your own model?
For tiny dataset, you mean a dataset with small images sizes (for example), or limited number of samples?

Are you able to try a different TPU node (remember to update the `XRT_TPU_CONFIG` environment variable with the new IP if you do that)?

**Comment by Crysitna** [comment]
I retry the code on nightly and 0.1, yet this time both are giving a new error:
```
StackTrace:
2019-08-06 08:28:57.253652: E tensorflow/compiler/xla/xla_client/xla_util.cc:72] *** Begin stack trace ***
2019-08-06 08:28:57.253657: E tensorflow/compiler/xla/xla_client/xla_util.cc:72]        tensorflow::CurrentStackTrace[abi:cxx11]()
2019-08-06 08:28:57.253663: E tensorflow/compiler/xla/xla_client/xla_util.cc:72]        xla::util::ReportComputationError(tensorflow::Status const&, absl::Span<xla::XlaComputation const* const>)
2019-08-06 08:28:57.253685: E tensorflow/compiler/xla/xla_client/xla_util.cc:72]        xla::util::ShapeHash(xla::Shape const&)
2019-08-06 08:28:57.253692: E tensorflow/compiler/xla/xla_client/xla_util.cc:72]        xla::XrtComputationClient::ExecuteComputation(xla::ComputationClient::Computation const&, absl::Span<std::shared_ptr<xla::ComputationClient::Data> const>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, xla::ComputationClient::ExecuteComputationOptions const&)

clone 

Status: Not found: From /job:tpu_worker/replica:0/task:0:
2019-08-06 08:28:57.253769: E tensorflow/compiler/xla/xla_client/xla_util.cc:72] XRT memory handle not found: 2098747143024929
2019-08-06 08:28:57.253774: E tensorflow/compiler/xla/xla_client/xla_util.cc:72]         [[{{node XRTExecute_128}}]]
Segmentation fault
```

single core wont produce the enqueue error, yet producing some other problem, and just like above, the problems can be different each time (session closed/ service not found/ etc)

the tiny dataset means limited data samples.

I'm afraid I have access to this tpu only..and I guess it wont be the hardware's problem since it works on the other model?

**Comment by dlibenzi** [comment]
Thanks @Crysitna .
So, to recap, MNIST works, and even your model with 1 core or a small dataset, works.
To be clear, your model works on 1 core with the full size dataset, right?
Did you check if your model with the small dataset train correctly (loss goes down according to what you experience on CPU/GPU - or similarly)?

It is hard to debug this w/out having a small repro.

**Comment by Crysitna** [comment]
all the models were migrated from the gpu code and has been extensively tested before - so should be good

for the recap, right now we hv 3 pairs model-dataset: 
1) mnist - mnist : works on both one core and 8 cores
2) model1 (small size) - small dataset :  runnable on both one core and 8 core _**I just realize the performance on 8 cores are quite bad while the results on one core is normal, yet I'm still figuring out the reason**_
3) model2 (large size) - relative large dataset: runnable on one core yet _**report the error on 8 cores**_

**Comment by dlibenzi** [comment]
Do you have any Python threading going on in your data loader?
Anything strange you do in your data loader?
How big is your dataset, in GB?

**Comment by Crysitna** [comment]
yeah I'm using _pescador_ to sampling from dataset and it does have a separate thread. code is like following: 
```
streamers = [pescador.Streamer(single_data_source_genf) for _ in ndatasource]    # prepare a separate streamer for each data source
mux = pescador.ShuffledMux(streamers)
dataset = PescadorDataset(mux)      # PescadorDataset: self-defined, inherit torch.utils.data.Dataset
dataloader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, num_workers=1, pin_memory=True)
```
so the document said pescador would issue a background process for data generation, and I monitored the thread, there r totally 2 threads running since the data loading part. No other strange things in data loader I guess.

the dataset is around 6GB

**Comment by dlibenzi** [comment]
Can you try to skip that data loader and feed synthetic data?
Similarly to what we do here:

https://github.com/pytorch/xla/blob/68704a54af5708acd011a3bd488478276616e160/test/test_train_mnist.py#L53

**Comment by tweetyone** [comment]
I met the same error: `Unable to enqueue when not opened`
I use the same data for different models. When I use simple model like resnet50, the code can run, but when I use complex model like senet154, it occurs `Unable to enqueue when not opened` error.  (The models I used is from https://github.com/Cadene/pretrained-models.pytorch) So I believe this error may have some relationship with high memory or long runtime.

**Comment by dlibenzi** [comment]
Is the above the only error message you get, or there is more?
Can we get the full error dump?

**Comment by tweetyone** [comment]
![image](https://user-images.githubusercontent.com/30562332/63004450-49a80800-be2f-11e9-8fc9-d0340238d8f2.png)

**Comment by dlibenzi** [comment]
This looks like an error one could get if we try to upload too much data onto the TPU device memory.
Can you try if a smaller batch size fixes the issue?

**Comment by tweetyone** [comment]
Tried a smaller batch size (8 and 4) which I can run on tesla p100 GPU before, but it still have the same error.

**Comment by dlibenzi** [comment]
This is on a TPU v2 or v3?

**Comment by tweetyone** [comment]
> This is on a TPU v2 or v3?

TPU v3

**Comment by dlibenzi** [comment]
How do you run your model?
Any small repro so we can try to debug?

**Comment by tweetyone** [comment]
> How do you run your model?
> Any small repro so we can try to debug?

https://github.com/tweetyone/DCL/blob/master/train_tpu.py

**Comment by dlibenzi** [comment]
> > How do you run your model?
> > Any small repro so we can try to debug?
> 
> https://github.com/tweetyone/DCL/blob/master/train_tpu.py

Hmm, the few lines starting from the line below look weird. Such tensors are already on TPU device.

https://github.com/tweetyone/DCL/blob/0168bd97c1812dbbb59d4b6dc5a8c67077006f85/train_tpu.py#L137

I'd probably create the losses outside of the loop here:

https://github.com/tweetyone/DCL/blob/0168bd97c1812dbbb59d4b6dc5a8c67077006f85/train_tpu.py#L163

**Comment by tweetyone** [comment]
Yep, thanks! The error still exist lol.
https://github.com/tweetyone/DCL/blob/0168bd97c1812dbbb59d4b6dc5a8c67077006f85/train_tpu.py#L137 
But for this part, I tested it before, these tensors are on cpu except for the first part of data[1]---labels on the tpu. I think it may be a small bug. reference: https://github.com/pytorch/xla/issues/899

**Comment by dlibenzi** [comment]
Posted a comment on #899 to verify the data type returned by your data loader.

Does the model run a few steps before giving error, or does it error out immediately?
Can you `print(torch_xla._XLAC._xla_metrics_report())` at each step?
